### PR TITLE
increase IO polling interval over time

### PR DESCRIPTION
**Pull request by martenson**
Introduces configurable option `--polling_backoff <integer>` which, atm, increases polling interval by `<integer>` second(s) every round. This considerably lowers the load on the target instance when waiting on dataset state changes.

xref: https://github.com/galaxyproject/usegalaxy-playbook/issues/164

### CODE REVIEW

File: docs/test_format.rst [hunk 1]
```diff
@@ -246,15 +246,16 @@ option to mount test data into the testing container.
...
-    $ planemo test --engine external_galaxy --galaxy_admin_key <admin_key> --galaxy_user_key <user_key> [--no_shed_install] <url>
+    $ planemo test --engine external_galaxy --galaxy_admin_key <admin_key> --galaxy_user_key <user_key> [--no_shed_install] [--easy_polling] <url>
 
 This is primarily useful for testing workflows against already running Galaxy instances. An admin or
 master API key should be supplied to install missing tool repositories for the workflow and a user API
 key should be supplied to run the workflow using. If you wish to skip tool shed repository installation
-(this requires the tool all be present already), use the ``-no_shed_install`` option with the ``test``
-command.
+(this requires all the tools be present already), use the ``--no_shed_install`` option. If you want to
+reduce the load on the target Galaxy while checking for the dtatus changes use the ``--easy_polling``
```
Comment (mvdbeek) [inline docs/test_format.rst]:
```suggestion
reduce the load on the target Galaxy while checking for the status changes use the ``--easy_polling``
```

File: docs/test_format.rst [hunk 2]
```diff
@@ -86,19 +86,19 @@ the above examples. Planemo adapts the CWL_ job document to Galaxy_ workflows an
 names for Galaxy_ tools and input node labels for workflows.
 
 Input files can be specified using either ``path`` attributes (which should generally be file
-paths relative to the aritfact and test directory) or ``location`` (which should be a URI). The 
+paths relative to the aritfact and test directory) or ``location`` (which should be a URI). The
```
Comment (nsoranzo) [inline docs/test_format.rst]: s/aritfact/artifact/

File: planemo/io.py [hunk 3]
```diff
@@ -285,20 +285,32 @@ def tee_captured_output(output):
...
+                raise Exception(message)
+            iteration += 1
+            value = function()
+            if value is not None:
+                return value
+            time.sleep(delta)
+    else:
+        timing = 0
+        while True:
+            if (timing) > timeout:
```
Comment (nsoranzo) [inline planemo/io.py]: Remove the parentheses?

File: planemo/io.py [hunk 4]
```diff
@@ -285,20 +285,32 @@ def tee_captured_output(output):
...
+                message = "Timed out waiting on %s." % desc
+                raise Exception(message)
+            next_sleep = delta + iteration
+            timing += next_sleep
+            iteration += 1
```
Comment (nsoranzo) [inline planemo/io.py]: You don't need `next_sleep` and `iteration`, you can replace lines 307-309 with:

```python
timing += delta
delta += polling_increase
```

File: planemo/io.py [hunk 5]
```diff
@@ -285,20 +285,32 @@ def tee_captured_output(output):
...
+            value = function()
+            if value is not None:
+                return value
+            time.sleep(next_sleep)
```
Comment (nsoranzo) [inline planemo/io.py]: time.sleep(delta)

File: planemo/io.py [hunk 6]
```diff
@@ -285,20 +285,32 @@ def tee_captured_output(output):
...
+        while True:
+            if (delta * iteration) > timeout:
+                message = "Timed out waiting on %s." % desc
+                raise Exception(message)
+            iteration += 1
+            value = function()
+            if value is not None:
+                return value
+            time.sleep(delta)
+    else:
```
Comment (nsoranzo) [inline planemo/io.py]: You can remove the whole `if` branch above by setting:

```python
polling_increase = int(use_easy_polling)
```

File: planemo/galaxy/activity.py [hunk 7]
```diff
@@ -638,8 +640,8 @@ def get_state():
             return state
         else:
             return None
-
-    final_state = wait_on(get_state, "state", timeout=60 * 60 * 24)
+    timeout=60 * 60 * 24
```
Comment (nsoranzo) [inline planemo/galaxy/activity.py]:
```suggestion
    timeout = 60 * 60 * 24
```

File: docs/test_format.rst [hunk 8]
```diff
@@ -246,15 +246,16 @@ option to mount test data into the testing container.
...
-    $ planemo test --engine external_galaxy --galaxy_admin_key <admin_key> --galaxy_user_key <user_key> [--no_shed_install] <url>
+    $ planemo test --engine external_galaxy --galaxy_admin_key <admin_key> --galaxy_user_key <user_key> [--no_shed_install] [--polling_backoff <integer>] <url>
 
 This is primarily useful for testing workflows against already running Galaxy instances. An admin or
 master API key should be supplied to install missing tool repositories for the workflow and a user API
 key should be supplied to run the workflow using. If you wish to skip tool shed repository installation
-(this requires the tool all be present already), use the ``-no_shed_install`` option with the ``test``
-command.
+(this requires all the tools be present already), use the ``--no_shed_install`` option. If you want to
+reduce the load on the target Galaxy while checking for the status changes use the ``--polling_backof <integer>``
```
Comment (nsoranzo) [inline docs/test_format.rst]:
```suggestion
reduce the load on the target Galaxy while checking for the status changes use the ``--polling_backoff <integer>``
```

**Comment by nsoranzo** [comment]
Thanks @martenson for changing it to an int, I wanted to suggest it, but than backed off :laughing:

**Pull request merged**